### PR TITLE
Remember all files in the clang-tidy cache that were clean.

### PR DESCRIPTION
**Pull request by hzeller**
Previously, we discarded the whole list of files when there was a single clang-tidy complaint in any of them.
Now, only exclude the files from the cache that had issues, but keep all the rest. That way it is possible to incrementally fix all the clang-tidy complatins without having to reprocess all the files that didn't have any issues.

clang-tidy is very expensive: running it over all verible files takes about 3-4 core hours. This will help reduce duplicate work.

### CODE REVIEW

File: .github/bin/run-clang-tidy.sh [hunk 1]
```diff
@@ -128,17 +129,35 @@ if [ -s ${FILES_TO_PROCESS} ]; then
...
+        /^[a-zA-Z_./]*:/ { seen[$1]=1; }
+        END {
+          is_first=1;
+          for (k in seen) {
+            printf("%s%s", is_first ? "/" : "\\|", gensub(/\//, "\\\\/", "g", k));
+            is_first=0;
+          }
+          printf("/d");
+        }' < ${TIDY_OUT})
+  sed "${EXCLUDE_EXPRESSION}" < ${CLANG_TIDY_SEEN_CACHE}.new > ${CLANG_TIDY_SEEN_CACHE}
```
Comment (fangism) [inline .github/bin/run-clang-tidy.sh:156]: `sed -e` (to be explicit)

Alternatively, you could send the output of `awk` to a file and use `sed -f` on that.  Might run less risk of hitting string length limits in the shell.

File: .github/bin/run-clang-tidy.sh [hunk 2]
```diff
@@ -128,17 +129,35 @@ if [ -s ${FILES_TO_PROCESS} ]; then
...
+  # in the clang-tidy output.
+  # Use that to filter the file list cache to leave out files with failures,
+  # but keep the ones that had no issues.
+  # That way, only files that had issues will have to be processed in
+  # subsequent clang-tidy runs.
+  EXCLUDE_EXPRESSION=$(awk -F: '
+        /^[a-zA-Z_./]*:/ { seen[$1]=1; }
+        END {
+          is_first=1;
+          for (k in seen) {
```
Comment (fangism) [inline .github/bin/run-clang-tidy.sh:150]: If there are concerns about ordering/stability, due to looping over a hashmap, consider sorting the output.

**Review by fangism — APPROVED** [review]

File: .github/bin/run-clang-tidy.sh [hunk 3]
```diff
@@ -128,17 +129,35 @@ if [ -s ${FILES_TO_PROCESS} ]; then
...
 else
   echo "Skipping clang-tidy run: nothing to do"
 fi
 
-# No complaints. We can cache this list now as baseline for next time.
-cp ${CLANG_TIDY_SEEN_CACHE}.new ${CLANG_TIDY_SEEN_CACHE}
 
-echo "No clang-tidy complaints.😎"
-exit 0
+if [ -s ${TIDY_OUT} ]; then
```
[the remark this answers is not in the document]
Comment (hzeller) [reply .github/bin/run-clang-tidy.sh]: Done here and partially elsewhere. Will do that separately.

File: .github/bin/run-clang-tidy.sh (hunk 1 above)
Comment (hzeller) [reply .github/bin/run-clang-tidy.sh]: Good idea. Done.

File: .github/bin/run-clang-tidy.sh (hunk 2 above)
Comment (hzeller) [reply .github/bin/run-clang-tidy.sh]: No concerns.

**Pull request merged**